File: core/src/main/java/io/github/msameer0/rhythmicrush/game/gameplay/interactables/portals/GravityPortal.java
package io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals;

import io.github.msameer0.rhythmicrush.game.registries.Registry;

/**
 * A portal that flips the player's gravity upon contact.
 */
@Registry(id = "gravity_portal")
public class GravityPortal extends AbstractPortal {

    private boolean flipped;

    /**
     * Constructs a new GravityPortal at the specified coordinates.
     *
     * @param x       the x-coordinate of the portal's position
     * @param y       the y-coordinate of the portal's position
     * @param flipped whether this portal flips gravity to upside down (true) or normal (false)
     */
    public GravityPortal(float x, float y, boolean flipped) {
        super(x, y);
        this.type = PortalType.GRAVITY;
        this.flipped = flipped;
    }

    /**
     * No-arg constructor for pooling.
     */
    public GravityPortal() {
        super();
        this.type = PortalType.GRAVITY;
    }

    public GravityPortal init(float x, float y, boolean flipped) {
        super.init(x, y);
        this.flipped = flipped;
        return this;
    }

    public boolean isFlipped() {
        return flipped;
    }
}

File: core/src/main/kotlin/io/github/msameer0/rhythmicrush/game/gameplay/interactables/portals/GravityPortal.kt
package io.github.msameer0.rhythmicrush.game.gameplay.interactables.portals

import io.github.msameer0.rhythmicrush.game.registries.Registry

/**
 * A portal that flips the direction of gravity for the player.
 */
@Registry(id = "gravity_portal")
class GravityPortal : AbstractPortal {
    constructor(x: Float, y: Float) : super(x, y) {
        this.type = PortalType.GRAVITY
        this.height = 200f
    }

    constructor() : super() {
        this.type = PortalType.GRAVITY
        this.height = 200f
    }

    override fun init(x: Float, y: Float, rotation: Float): GravityPortal {
        super.init(x, y, rotation)
        this.type = PortalType.GRAVITY
        this.height = 200f
        return this
    }

    override fun init(x: Float, y: Float): GravityPortal {
        return init(x, y, 0f)
    }
}
